### ISSUE TITLE
[no-unnecessary-type-assertion] False positive with flow-based narrowing

### ISSUE THREAD

**Issue by taymonbeal** [typescript-eslint/typescript-eslint#3310]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "plugins": ["@typescript-eslint"],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "sourceType": "module",
    "project": ["./tsconfig.json"]
  },
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": ["error"]
  }
}
```

```TS
export function nonNullish<T>(value: T): NonNullable<T> {
  if (value == null) {
    throw new TypeError();
  }
  return value!;
}
```

```JSON
{ "compilerOptions": { "strict": true } }
```

**Expected Result**

No errors. (Note that removing the `!` causes TypeScript to complain.)

**Actual Result**

```
  5:10  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
```

**Additional Info**

I think this isn't a duplicate of #2248 because I checked on ts-ast-viewer and `any` didn't show up anywhere.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.1-alpha.11` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.25.0` |
| `node`                             | `12.21.0` |

**Comment by bradzacher** [comment]
The issue here is that `typeof value === T` - a generic type.
A generic type is "not a nullish type" - it is just an amorphous blob thing.
Which our code doesn't know how to handle.

---

Another way to implement this is as follows, which no longer requires the non-null assertion.
```ts
function nonNullish<T>(value: T | null | undefined): T {
  if (value == null) {
    throw new TypeError();
  }
  return value;
}
```

**Comment by Josh-Cena** [comment]
I have [a similar case](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=GYVwdgxgLglg9mABMOcA8AVRBTAHlbMAEwGdESoAnGMAc0QB9EwQBbAI20oD4AKANwCGAGwBciDAEpEAbwBQiRBAQUlIiCGGCCRRAF5EUAJ4AHbHGCIhw-XoMAiCtTr3EAfkQAGROIEjEgmQsHFySANwKiJTYUCCUSBDqmtrYugCEdojgRNjANKnuXgB0AGyIAFRqwhpaOj5eEQC+QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA) which I think is in the same vein as this one.

```ts
function foo<T extends string | number>(val: T) {
  const calculated = typeof val === "string" ? 0: (val as number);
  return calculated !== undefined ? 0.6 * calculated : 0;
}
```

The `val` type is `T`, and it's not subject to control flow narrowing, so if I remove the `as number`, `0.6 * calculated` will fail.

**Comment by taymonbeal** [comment]
I don't think it's correct to say that this issue applies only to generic types. See, e.g., https://typescript-eslint.io/play/#ts=5.2.2&fileType=.ts&code=GYVwdgxgLglg9mABAIwIYCcAUATAXIgEVSgFMBKRAbwChFEIEBnOAGxIDoW4BzHMgbmoBfatVCRYCRMDhxMFGnTZREAD0QBeRGBAsWgujGCJMAWWIALdulRhscALbzEAHkQAGdgFYFtOms1tEgB3QmISeQNEETo0LFUAQgFhaiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFuQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

**Comment by bradzacher** [comment]
@taymonbeal that is a different issue entirely that is actually a regression in TS itself
https://github.com/typescript-eslint/typescript-eslint/issues/10334 / https://github.com/microsoft/TypeScript/issues/60514
